File: backend/app/ecommerce/schemas/product.py
"""Product schemas for API requests and responses."""

from decimal import Decimal
from typing import List, Optional
from datetime import datetime

from pydantic import BaseModel, HttpUrl


class ProductCreate(BaseModel):
    """Schema for creating a new product."""
    
    name: str
    url: HttpUrl
    site: str
    category: Optional[str] = None


class ProductUpdate(BaseModel):
    """Schema for updating a product."""
    
    name: Optional[str] = None
    category: Optional[str] = None
    is_tracked: Optional[bool] = None


class PriceHistoryResponse(BaseModel):
    """Schema for price history response."""
    
    id: int
    price: Decimal
    currency: str
    availability: Optional[str]
    created_at: datetime
    
    class Config:
        from_attributes = True


class ProductResponse(BaseModel):
    """Schema for product response."""
    
    id: int
    name: str
    url: str
    site: str
    category: Optional[str]
    is_tracked: bool
    is_active: bool
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class ProductDetailResponse(ProductResponse):
    """Schema for detailed product response with price history."""
    
    price_history: List[PriceHistoryResponse] = []


class ProductListResponse(BaseModel):
    """Schema for paginated product list response."""
    
    items: List[ProductResponse]
    total: int
    page: int
    size: int
    pages: int

File: frontend/src/types/product.ts
export interface Product {
  id: number;
  name: string;
  url: string;
  site: string;
  category?: string;
  is_tracked: boolean;
  created_at: string;
  updated_at: string;
}

export interface PriceHistory {
  id: number;
  product_id: number;
  price: number;
  currency: string;
  created_at: string;
}

export interface CreateProductPayload {
  name: string;
  url: string;
  category?: string;
}

export interface ProductStats {
  total_count: number;
  active_deals_count: number;
  price_drops_week: number;
}